### DISCUSSION TITLE
Why won't a correlation matrix plot render when formatting with Typst?

### DISCUSSION THREAD

**Discussion by baker-jr-john** [quarto-dev/quarto-cli#9217] [Q&A] [answered]
### Description

Hi! 👋 I'm trying to render a correlation matrix from a .qmd file to .pdf using the Typst format in RStudio. The plot displays as expected in RStudio, but only a black rectangle is in the .pdf.
<img width="1552" alt="Screenshot 2024-03-28 at 1 36 49 PM" src="https://github.com/quarto-dev/quarto-cli/assets/71906203/49789b5a-4edc-4aa1-be79-becfc87d9222">
<img width="1440" alt="Screenshot 2024-03-28 at 1 39 14 PM" src="https://github.com/quarto-dev/quarto-cli/assets/71906203/928b9bd1-b621-4d4c-88a6-f0b2db7b9dd2">

I'm using Quarto 1.4.550 with macOS 14.4.1. Here's a reproducible example:

````qmd
---
title: "Impact of House Features on Sale Prices"
author:   
  - John Baker
format:
  typst:
    toc: true
    number-sections: true
---
```{r}
# Load libraries
library(ggplot2)
library(ggcorrplot)

# Load vectors
Id <- c(1, 2, 3, 4, 5)
MSSubClass <- c(60, 20, 60, 70, 60)
MSZoning <- as.factor(c("RL", "RL", "RL", "RL", "RL"))
LotFrontage <-
  c(65, 80, 68, 60, NA) # Assuming NA for missing values
LotArea <- c(8450, 9600, 11250, 9550, 14260)
Street <- as.factor(c("Pave", "Pave", "Pave", "Pave", "Pave"))

# Combine into a data frame
df <-
  data.frame(Id, MSSubClass, MSZoning, LotFrontage, LotArea, Street)

# Select only numeric columns from train_data
numeric_cols <- sapply(df, is.numeric)
train_numeric <- df[, numeric_cols]

# Calculate the correlation matrix
corrmat <- cor(train_numeric, use = "pairwise.complete.obs")

# Create the correlation plot using ggcorrplot
ggcorrplot(corrmat,
           hc.order = TRUE,
           tl.cex = 5)
```
````
I appreciate your help!

**Comment by cscheid** [comment]
Can you provide an example that doesn't need that train.csv file? Thanks.

**Reply by baker-jr-john** [reply]
I updated my example.

**Reply by mcanouil** [reply, marked as answer]
The issue is the format of the figure.
If you set `fig-format: png` it works.

---

You can look at the intermediate markdown and typst file using the `keep-*` options:

````
---
title: "Impact of House Features on Sale Prices"
author:   
  - John Baker
format:
  typst:
    toc: true
    number-sections: true
    keep-md: true
    keep-typ: true
---
````

**Reply by cscheid** [reply]
I think this is a limitation on Typst's SVG renderer. Typst attempts to render SVG directly into the PDF and the SVG being emitted is not being handled correctly.

**Reply by baker-jr-john** [reply]
I was hoping it was something else. It doesn't look so good as a .png, but it works. I appreciate it! I hope they fix this.

**Reply by cscheid** [reply]
@baker-jr-john You'll need to either take the resulting .typ file with the SVG and file a bug at https://github.com/typst/typst or use a different rendering format as @mcanouil suggested. Cf https://typst.app/docs/reference/visualize/image/:

> Note: Work on SVG export is ongoing and there might be visual inaccuracies in the resulting PDF. Make sure to double-check embedded SVG images. If you have an issue, also feel free to report it on [GitHub](https://github.com/typst/typst/issues?q=is%3Aopen+is%3Aissue+label%3Asvg).

From inspecting the .svg file, I suspect the issue is that this file uses SVG filters. Using plain ggplot() calls produces a file that works with Typst and doesn't have SVG filters like these:

```
<filter id="filter-remove-color-and-invert-alpha" x="0%" y="0%" width="100%" height="100%">
<feColorMatrix color-interpolation-filters="sRGB" values="0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 -1 1"/>
</filter>
<filter id="filter-0" x="0%" y="0%" width="100%" height="100%">
<feImage xlink:href="#compositing-group-1" result="source" x="0" y="0" width="475.2" height="302.4"/>
<feImage xlink:href="#compositing-group-2" result="destination" x="0" y="0" width="475.2" height="302.4"/>
<feComposite in="source" in2="destination" operator="arithmetic" k1="0" k2="1" k3="1" k4="0" color-interpolation-filters="sRGB"/>
</filter>
```

**Reply by cscheid** [reply]
It looks like this definition is ultimately coming from Cairo's SVG device: https://github.com/search?type=code&auto_enroll=true&q=filter-remove-color-and-invert-alpha+language%3AC&l=C

**Reply by mcanouil** [reply]
Then I suggest to use `svglite`:

````qmd
---
title: "Impact of House Features on Sale Prices"
author:   
  - John Baker
format:
  typst:
    toc: true
    number-sections: true
knitr:
  opts_chunk:
    dev: svglite
---

```{r}
# Load libraries
library(ggplot2)
library(ggcorrplot)

# Load vectors
Id <- c(1, 2, 3, 4, 5)
MSSubClass <- c(60, 20, 60, 70, 60)
MSZoning <- as.factor(c("RL", "RL", "RL", "RL", "RL"))
LotFrontage <-
  c(65, 80, 68, 60, NA) # Assuming NA for missing values
LotArea <- c(8450, 9600, 11250, 9550, 14260)
Street <- as.factor(c("Pave", "Pave", "Pave", "Pave", "Pave"))

# Combine into a data frame
df <-
  data.frame(Id, MSSubClass, MSZoning, LotFrontage, LotArea, Street)

# Select only numeric columns from train_data
numeric_cols <- sapply(df, is.numeric)
train_numeric <- df[, numeric_cols]

# Calculate the correlation matrix
corrmat <- cor(train_numeric, use = "pairwise.complete.obs")

# Create the correlation plot using ggcorrplot
ggcorrplot(
  corrmat,
  hc.order = TRUE,
  tl.cex = 5
)
```
````

**Reply by cscheid** [reply]
Very interesting! (And the SVG that svglite generates doesn't actually emit filters, so I think my suspicion is correct)

I wonder if we should warn users that render Typst documents in R with the Cairo device...

**Reply by mcanouil** [reply]
I personally only uses [`svglite`](https://svglite.r-lib.org/) and `ragg_png` from [`ragg`](https://ragg.r-lib.org/) for all figures in R.
Those devices have the `scaling` parameter which makes life so much easier.
